src/persist-client/src/internal/maintenance.rs

```diff
@@ -242,38 +252,47 @@ where
         V: Debug + Codec,
         D: Semigroup + Codec64 + Send + Sync,
     {
-        for future in self.perform_in_background(machine, gc, compactor) {
-            let _ = future.await;
-        }
+        self.perform_in_background(machine, gc, compactor).await
     }
 
     /// Initiates maintenance work in the background, either through spawned tasks
     /// or by sending messages to existing tasks. The returned futures may be
     /// awaited to know when the work is completed, but do not need to be polled
     /// to drive the work to completion.
-    fn perform_in_background<K, V, D>(
+    async fn perform_in_background<K, V, D>(
         self,
         machine: &Machine<K, V, T, D>,
         gc: &GarbageCollector<K, V, T, D>,
-        compactor: Option<&Compactor<K, V, T, D>>,
-    ) -> Vec<BoxFuture<'static, ()>>
-    where
+        mut compactor: Option<&Compactor<K, V, T, D>>,
+    ) where
         K: Debug + Codec,
         V: Debug + Codec,
         D: Semigroup + Codec64 + Send + Sync,
     {
-        let mut futures = self.routine.perform_in_background(machine, gc);
+        let Self {
+            routine,
+            compaction,
+        } = self;
+        let mut more_maintenance = RoutineMaintenance::default();
```

Comment:
huge nit here: feels a bit more idiomatic to me to make this an option instead of the mem::take stuff, and then only set it to None if we get back a non-empty maintenance. then below make the while loop a `while let Some(...) {` and do the same thing

Reply:
I... think I disagree! In particular, we need to be very careful to avoid setting `more_maintenance` to `Some(default)` or we'd have an infinite loop, and the assignment code gets a little more gnarly:

```rust
let maintenance = future.await;
if !maintenance.is_empty() {
  match &mut more_maintenance {
    Some(m) => m.merge(maintenance),
    None => *m = Some(maintenance),
  }
}
```

I may totally be missing some nicer option combinator for this, but it feels fragile, and we'd have to make sure to do that every time we update `more_maintenance`.

(Totally happy to hear arguments though if you'd like to elaborate!)

Reply:
ah, went to type up concretely what I meant and the bit I was missing was that `routine.perform_in_background` returns more than one, which makes my suggestion far more clunky. retracting my suggestion!

(for posterity, reasoning was mostly just that I find `mem:take` is sometimes totally correct, but often a smell. especially when called on an option)